components/brave_shields/browser/domain_block_navigation_throttle.cc

```diff
@@ -0,0 +1,165 @@
+
+#include <memory>
+#include <utility>
+
+#include "base/bind.h"
+#include "base/task/post_task.h"
+#include "base/threading/thread_task_runner_handle.h"
+#include "brave/components/brave_shields/browser/ad_block_custom_filters_service.h"
+#include "brave/components/brave_shields/browser/ad_block_service.h"
+#include "brave/components/brave_shields/browser/domain_block_controller_client.h"
+#include "brave/components/brave_shields/browser/domain_block_page.h"
+#include "brave/components/brave_shields/browser/domain_block_tab_storage.h"
+#include "components/prefs/pref_service.h"
+#include "components/security_interstitials/content/security_interstitial_tab_helper.h"
+#include "components/user_prefs/user_prefs.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/browser_thread.h"
+#include "content/public/browser/navigation_handle.h"
+#include "content/public/browser/web_contents.h"
+#include "content/public/browser/web_contents_user_data.h"
+#include "net/base/net_errors.h"
+
+namespace {
+
+bool ShouldBlockDomainOnTaskRunner(
+    brave_shields::AdBlockService* ad_block_service,
+    const GURL& url) {
+  if (!ad_block_service->IsInitialized())
+    return false;
+  bool did_match_exception = false;
+  bool did_match_rule = false;
+  bool did_match_important = false;
+  std::string mock_data_url;
+  ad_block_service->ShouldStartRequest(
+      url, blink::mojom::ResourceType::kMainFrame, url.host(), &did_match_rule,
+      &did_match_exception, &did_match_important, &mock_data_url);
+  return (did_match_important || (did_match_rule && !did_match_exception));
+}
+
+}  // namespace
+
+namespace brave_shields {
+
+// static
+std::unique_ptr<DomainBlockNavigationThrottle>
+DomainBlockNavigationThrottle::MaybeCreateThrottleFor(
+    content::NavigationHandle* navigation_handle,
+    AdBlockService* ad_block_service,
+    AdBlockCustomFiltersService* ad_block_custom_filters_service,
+    const std::string& locale) {
+  if (!ad_block_service || !ad_block_custom_filters_service)
+    return nullptr;
+  return std::make_unique<DomainBlockNavigationThrottle>(
+      navigation_handle, ad_block_service, ad_block_custom_filters_service,
+      locale);
+}
+
+DomainBlockNavigationThrottle::DomainBlockNavigationThrottle(
+    content::NavigationHandle* navigation_handle,
+    AdBlockService* ad_block_service,
+    AdBlockCustomFiltersService* ad_block_custom_filters_service,
+    const std::string& locale)
+    : content::NavigationThrottle(navigation_handle),
+      ad_block_service_(ad_block_service),
+      ad_block_custom_filters_service_(ad_block_custom_filters_service),
+      locale_(locale) {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+}
+
+DomainBlockNavigationThrottle::~DomainBlockNavigationThrottle() {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+}
+
+content::NavigationThrottle::ThrottleCheckResult
+DomainBlockNavigationThrottle::WillStartRequest() {
```

Comment:
is it called for redirects? We probably need a separate check in `WillRedirectRequest()`

Reply:
added in https://github.com/brave/brave-core/commit/1d30188eb955ce9355f7426f5e603563abccf567